<a href="https://colab.research.google.com/github/data-srkh/data-analytic-internship/blob/main/Task1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [70]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import zscore

#Making dataframes

In [71]:
purchase_data = pd.read_csv('QVI_purchase_behaviour.csv')
purchase_data.head()

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0            1000   YOUNG SINGLES/COUPLES          Premium
1            1002   YOUNG SINGLES/COUPLES       Mainstream
2            1003          YOUNG FAMILIES           Budget
3            1004   OLDER SINGLES/COUPLES       Mainstream
4            1005  MIDAGE SINGLES/COUPLES       Mainstream

In [72]:
transaction_data = pd.read_excel('QVI_transaction_data.xlsx')
transaction_data.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8

#Creating Summaries

In [73]:
purchase_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72637 entries, 0 to 72636
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   LYLTY_CARD_NBR    72637 non-null  int64 
 1   LIFESTAGE         72637 non-null  object
 2   PREMIUM_CUSTOMER  72637 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.7+ MB


In [74]:
purchase_data.describe()

LYLTY_CARD_NBR
count    7.263700e+04
mean     1.361859e+05
std      8.989293e+04
min      1.000000e+03
25%      6.620200e+04
50%      1.340400e+05
75%      2.033750e+05
max      2.373711e+06

Summary: The dataset has 2 categorical and 1 numerical column.  There are 72,637 values for LYLTY_CARD_NBR with mean 136,185.9.

In [75]:
transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 264836 entries, 0 to 264835
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   DATE            264836 non-null  int64  
 1   STORE_NBR       264836 non-null  int64  
 2   LYLTY_CARD_NBR  264836 non-null  int64  
 3   TXN_ID          264836 non-null  int64  
 4   PROD_NBR        264836 non-null  int64  
 5   PROD_NAME       264836 non-null  object 
 6   PROD_QTY        264836 non-null  int64  
 7   TOT_SALES       264836 non-null  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 16.2+ MB


In [76]:
transaction_data.describe()

DATE     STORE_NBR  LYLTY_CARD_NBR        TXN_ID  \
count  264836.000000  264836.00000    2.648360e+05  2.648360e+05   
mean    43464.036260     135.08011    1.355495e+05  1.351583e+05   
std       105.389282      76.78418    8.057998e+04  7.813303e+04   
min     43282.000000       1.00000    1.000000e+03  1.000000e+00   
25%     43373.000000      70.00000    7.002100e+04  6.760150e+04   
50%     43464.000000     130.00000    1.303575e+05  1.351375e+05   
75%     43555.000000     203.00000    2.030942e+05  2.027012e+05   
max     43646.000000     272.00000    2.373711e+06  2.415841e+06   

            PROD_NBR       PROD_QTY      TOT_SALES  
count  264836.000000  264836.000000  264836.000000  
mean       56.583157       1.907309       7.304200  
std        32.826638       0.643654       3.083226  
min         1.000000       1.000000       1.500000  
25%        28.000000       2.000000       5.400000  
50%        56.000000       2.000000       7.400000  
75%        85.000000       2.000000       9.200000  
max       114.000000     200.000000     650.000000

Summary: There are 7 numerical and 1 categorical columns.


#Finding and removing outliers

I used Z score method to detect outliers.


In [77]:
z_scores = np.abs(zscore(purchase_data.select_dtypes(include=[np.number])))
threshold = 3
outliers = (z_scores > threshold).any(axis=1)
print("Outliers detected using Z-score:")
print(purchase_data[outliers])

Outliers detected using Z-score:
       LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
72603          861921   OLDER SINGLES/COUPLES          Premium
72604          861951                RETIREES           Budget
72605          861961          YOUNG FAMILIES           Budget
72606          862501          YOUNG FAMILIES           Budget
72607          880171          OLDER FAMILIES           Budget
72608          880551   OLDER SINGLES/COUPLES          Premium
72609          880711          OLDER FAMILIES           Budget
72610          883791   OLDER SINGLES/COUPLES       Mainstream
72611         2330031  MIDAGE SINGLES/COUPLES          Premium
72612         2330041   YOUNG SINGLES/COUPLES       Mainstream
72613         2330051                RETIREES          Premium
72614         2330081                RETIREES       Mainstream
72615         2330121                RETIREES           Budget
72616         2330171                RETIREES       Mainstream
72617         2330191 

In [78]:
purchase_df = purchase_data[~outliers]
print("DataFrame after removing outliers:")
print(purchase_df)

DataFrame after removing outliers:
       LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0                1000   YOUNG SINGLES/COUPLES          Premium
1                1002   YOUNG SINGLES/COUPLES       Mainstream
2                1003          YOUNG FAMILIES           Budget
3                1004   OLDER SINGLES/COUPLES       Mainstream
4                1005  MIDAGE SINGLES/COUPLES       Mainstream
...               ...                     ...              ...
72598          272388   OLDER SINGLES/COUPLES          Premium
72599          272389   YOUNG SINGLES/COUPLES       Mainstream
72600          272390            NEW FAMILIES           Budget
72601          272391   YOUNG SINGLES/COUPLES       Mainstream
72602          272392  MIDAGE SINGLES/COUPLES          Premium

[72603 rows x 3 columns]


In [79]:
z_scores = np.abs(zscore(transaction_data.select_dtypes(include=[np.number])))
threshold = 3
outliers = (z_scores > threshold).any(axis=1)
print("Outliers detected using Z-score:")
print(transaction_data[outliers])

Outliers detected using Z-score:
         DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
3       43329          2            2373     974        69   
11      43332          8            8294    8221       114   
16      43330         20           20418   17413        94   
31      43600         43           43227   40186        26   
56      43601         74           74336   73182        84   
...       ...        ...             ...     ...       ...   
258721  43327        200          200248  199694         3   
258726  43332        203          203253  203360        28   
258729  43601        208          208205  207318        37   
258759  43599        238          238169  242560        44   
258788  43599        264          264149  262909        25   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
3         Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
11         Kettle Sensations   Siracha Lime 150g         5       23.0  
16    

In [80]:
transaction_df = transaction_data[~outliers]
print("DataFrame after removing outliers:")
print(transaction_df)

DataFrame after removing outliers:
         DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0       43390          1            1000       1         5   
1       43599          1            1307     348        66   
2       43605          1            1343     383        61   
4       43330          2            2426    1038       108   
5       43604          4            4074    2982        57   
...       ...        ...             ...     ...       ...   
264831  43533        272          272319  270088        89   
264832  43325        272          272358  270154        74   
264833  43410        272          272379  270187        51   
264834  43461        272          272379  270188        42   
264835  43365        272          272380  270189        74   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
0         Natural Chip        Compny SeaSalt175g         2        6.0  
1                       CCs Nacho Cheese    175g         3        6.3  
2   

#Checking and correcting data formats

In [81]:
purchase_df['LYLTY_CARD_NBR'] = pd.to_numeric(purchase_df['LYLTY_CARD_NBR'], errors='coerce')

<ipython-input-81-e88d46448e55>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  purchase_df['LYLTY_CARD_NBR'] = pd.to_numeric(purchase_df['LYLTY_CARD_NBR'], errors='coerce')


In [82]:
transaction_df['STORE_NBR'] = pd.to_numeric(transaction_df['STORE_NBR'], errors='coerce')
transaction_df['LYLTY_CARD_NBR'] = pd.to_numeric(transaction_df['LYLTY_CARD_NBR'], errors='coerce')
transaction_df['TXN_ID'] = pd.to_numeric(transaction_df['TXN_ID'], errors='coerce')
transaction_df['PROD_NBR'] = pd.to_numeric(transaction_df['PROD_NBR'], errors='coerce')
transaction_df['PROD_QTY'] = pd.to_numeric(transaction_df['PROD_QTY'], errors='coerce')
transaction_df['TOT_SALES'] = pd.to_numeric(transaction_df['TOT_SALES'], errors='coerce')

<ipython-input-82-887690f1975e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  transaction_df['STORE_NBR'] = pd.to_numeric(transaction_df['STORE_NBR'], errors='coerce')
<ipython-input-82-887690f1975e>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  transaction_df['LYLTY_CARD_NBR'] = pd.to_numeric(transaction_df['LYLTY_CARD_NBR'], errors='coerce')
<ipython-input-82-887690f1975e>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_in

In [83]:
transaction_df['DATE'] = pd.to_datetime(transaction_df['DATE'], format='%Y-%m-%d', errors='coerce')

<ipython-input-83-c0b40fce3d35>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  transaction_df['DATE'] = pd.to_datetime(transaction_df['DATE'], format='%Y-%m-%d', errors='coerce')
